gitdiff/combined.go

```diff
@@ -0,0 +1,269 @@
+		}
+
+		hasAdd := false
+		hasRemove := false
+		hasContext := false
+		for idx, op := range parentOps {
+			frag := frags[idx]
+			altered[idx] = false
+
+			switch op {
+			case ' ':
+				// Context lines
+				hasContext = true
+				frag.Lines = append(frag.Lines, Line{OpContext, data})
+				// Adjustment of the leading and trailing context count can only happen
+				// by analyzing all the file operations, so that happens after the line's
+				// operation checks.
+			case '-':
+				hasRemove = true
+				altered[idx] = true
+				oldLines--
+				noLineChanges = false
+				frag.LinesDeleted++
+				trailingContext = 0
+				frag.Lines = append(frag.Lines, Line{OpDelete, data})
+			case '+':
+				hasAdd = true
+				altered[idx] = true
+				noLineChanges = false
+				frag.LinesAdded++
+				trailingContext = 0
+				frag.Lines = append(frag.Lines, Line{OpAdd, data})
+			case '\\':
+				// this may appear in middle of fragment if it's for a deleted line
+				if isNoNewlineMarker(line) {
+					removeLastNewline(frag)
+					// Move on to the next line.
+					continue lineLoop
+				}
+				fallthrough
+			default:
+				// TODO(bkeyes): if this is because we hit the next header, it
+				// would be helpful to return the miscounts line error. We could
+				// either test for the common headers ("@@ -", "diff --git", "@@@ -") or
+				// assume any invalid op ends the fragment; git returns the same
+				// generic error in all cases so either is compatible
+				return p.Errorf(0, "invalid line operation: %q", op)
+			}
+		}
+
+		// The complex counting method.
+
+		if hasAdd {
+			// If any number of parent files is marked as an add, then count that
+			// as a single add for the new count.
+			newLines--
+		}
+
+		// Lines with removes reduce the old line count once per removal operation, and
+		//   the counting happens during each file's removal action.
+		if !hasRemove {
+			// For lines that have no removes, then this is either an add or a
+			// context line.  In both cases, the files which had a blank operation count
+			// as an old line.
+			for _, fragChanged := range altered {
+				if !fragChanged {
+					oldLines--
+				}
+			}
+			if !hasAdd && hasContext {
+				// Lines with no removes, no adds, and had at least 1 context entry
+				// means that this line a full context - no add and no remove.
+				if noLineChanges {
+					leadingContext++
+				} else {
+					trailingContext++
+				}
+			}
+		}
+
+		if err := p.Next(); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+	}
+
+	if oldLines != 0 || newLines != 0 {
+		hdr := max(allOldLines-oldLines, frags[0].NewLines-newLines) + 1
+		return p.Errorf(-hdr, "fragment header miscounts lines: %+d old, %+d new", -oldLines, -newLines)
+	}
+	if noLineChanges {
+		return p.Errorf(0, "fragment contains no changes")
+	}
+
+	// check for a final "no newline" marker since it is not included in the
+	// counters used to stop the loop above
+	if isNoNewlineMarker(p.Line(0)) {
```

Comment:
This reuses functionality from `text.go`; some future refactoring could centralize these functions as well as reduce the freshly-added code duplication.